### ISSUE TITLE
Bug(CD): Temperatures Displayed in Fahrenheit Instead of Celsius

### ISSUE THREAD

**Issue by hexCut** [wuwentao/midea_ac_lan#420]
### Current integration version

0.4.3

### Last known working integration version

0.4.3

### Device type and model

Heat Pump Water Heater RSJ000CB (9749) by Midea

### Working mobile app

SmartHome

### The detail description of bug/problem

Description

The integration for the Swan 2 Water Pump appears to assume that the device’s temperature readings are provided in Celsius. However, the device actually returns temperatures in Fahrenheit, which are displayed incorrectly in Home Assistant without conversion.

This discrepancy affects both:
	•	The current temperature (current_temperature property)
	•	The target temperature (target_temperature property)

Expected Behavior

Temperatures should be displayed in Celsius, as assumed by the integration.

Observed Behavior

Temperatures are displayed in Fahrenheit but treated as if they were Celsius, resulting in incorrect readings and user confusion.

Steps to Reproduce
	1.	Integrate the Swan 2 Water Pump using the current version of the integration.
	2.	Observe the current_temperature and target_temperature values in Home Assistant.
	3.	Compare with the actual Fahrenheit readings from the device.

### The logs

2024-12-15 13:01:37.830 DEBUG (Clivet Swan 2) [midealocal.device] [153931628204291] refresh_status with cmds: [<midealocal.devices.cd.message.MessageQuery object at 0xffff700a70e0>], check_protocol False
2024-12-15 13:01:37.831 DEBUG (Clivet Swan 2) [midealocal.device] [153931628204291] Sending: {'_device_type': <DeviceType.CD: 205>, '_message_type': <MessageType.query: 3>, '_body_type': <ListTypes.X01: 1>, '_message_protocol_version': 0, 'header': 'aa0ccd00000000000003', 'body': '0101', 'message_type': 'query', 'body_type': '01', 'self': <midealocal.devices.cd.message.MessageQuery object at 0xffff700a70e0>}, query is True
2024-12-15 13:01:38.505 DEBUG (Clivet Swan 2) [midealocal.devices.cd] [153931628204291] Received: {'_device_type': <DeviceType.CD: 205>, '_message_type': <MessageType.query: 3>, '_body_type': <ListTypes.X01: 1>, '_message_protocol_version': 0, '_header': 'aa3ecd00000000000003', '_body': <midealocal.devices.cd.message.CDGeneralMessageBody object at 0xffff74af8050>, '_data': '01010534307f7f2a441f4626090511001105083200340000000000000703000000000400000000000000000000060b000000b300', 'parser_list': [], 'power': True, 'target_temperature': 11, 'mode': 1, 'current_temperature': 9, 'condenser_temperature': 6.0, 'outdoor_temperature': 19.0, 'compressor_temperature': 0.5, 'max_temperature': 20, 'min_temperature': 4, 'compressor_status': False, 'header': 'aa3ecd00000000000003', 'body': '01010534307f7f2a441f4626090511001105083200340000000000000703000000000400000000000000000000060b000000b300', 'message_type': 'query', 'body_type': '01', 'self': <midealocal.devices.cd.message.MessageCDResponse object at 0xffff7042b3e0>}
2024-12-15 13:01:38.506 DEBUG (Clivet Swan 2) [midealocal.device] [153931628204291] Status update: {'power': True, 'mode': 'Standard', 'max_temperature': 20, 'min_temperature': 4, 'target_temperature': 11, 'current_temperature': 9, 'outdoor_temperature': 19.0, 'condenser_temperature': 6.0, 'compressor_temperature': 0.5, 'compressor_status': False}
![Uploading Screenshot 2024-12-15 at 13.02.10.png…]()

**Comment by wuwentao** [comment]
@hexCut all the temperature is Celsius in current midea_ac_lan. 
please don't compare it with Fahrenheit, it's a error action.
if you want to display Celsius to Fahrenheit, just update your HA setting. it's not midea_ac_lan feature.
so I will close current issue. and you can check how to update your HA setting from Celsius to Fahrenheit in HA web UI.

in addtion, if your device result is error, please description the detail error and attach debug log and app screenshot or device screenshot. and reopen current issue as it still have bug.
but you should use the same unit, and not do this compare.
thanks

**Comment by hexCut** [comment]
Thank you for your earlier response. However, I have conducted additional testing and gathered logs that clearly show the device (Clivet Swan 2) with the integration reports temperature values in Fahrenheit, not Celsius.

Below is an excerpt from the debug log:

2024-12-15 13:01:38.506 DEBUG (Clivet Swan 2) [midealocal.device] [153931628204291] Status update: {
    'power': True,
    'mode': 'Standard',
    'max_temperature': 20,
    'min_temperature': 4,
    'target_temperature': 11,
    'current_temperature': 9,
    'outdoor_temperature': 19.0,
    'condenser_temperature': 6.0,
    'compressor_temperature': 0.5,
    'compressor_status': False
}

	•	The device’s current_temperature is reported as 9°F, which corresponds to approximately -12.8°C.
(This is clearly not in Celsius, as 9°C would be a normal room temperature, not suitable for the device’s function.)
	•	Similarly, the target_temperature of 11°F corresponds to approximately -11.7°C, which is also not realistic in Celsius.

The integration should detect that the device reports temperatures in Fahrenheit and convert them to Celsius before displaying them in Home Assistant.

**Comment by wuwentao** [comment]
thanks, there is a bug for CD device, also reported in this github issue:
https://github.com/wuwentao/midea_ac_lan/issues/347

could you help to confirm with it ? 
in addtion, is possible for you to share your device SN to me ?

I will try to add a solution for https://github.com/wuwentao/midea_ac_lan/issues/347, and want to confirm with your device SN.

**Comment by wuwentao** [comment]
I'm not sure with this SN.
it should be a along Str.

**Comment by hexCut** [comment]
No, actually that serial number is taken directly from a label on the water heater itself. Where should I retrieve the correct serial number from? I couldn’t find anything related to it in the app. Could you clarify?

Thank you!

**Comment by wuwentao** [comment]
@hexCut yes, I can understand it, midea have more differnt types SN, used for different purpose.

1. below is a simple step: 
just backup your device xxx.json or rename it.
remove your device, enable debug log, add it with discovery, you can found the SN info in debug log.
for example:

```
2024-12-15 00:26:55.828 DEBUG (MainThread) [midealocal.discover] Found a supported device: {'device_id': xxx, 'type': 172, 'ip_address': '192.168.1.22', 'port': 6444, 'model': 'xxx', 'sn': 'xxx', 'protocol': 3}
```

this SN is what we expected.


method 3:  in addition, if you can login to HAOS and start docker.
also can be do below manual steps:

1. ssh login to HAOS
2. run docker with `docker exec -it homeassistant /bin/bash`
3. run discover debug command: `python3 -m midealocal.cli discover -d --host x.x.x.x` (x.x.x.x is your device IP)

you can also found the same output as debug log.

we can use your device SN to confirm your device protocol details.

**Comment by hexCut** [comment]
Ok, can it be this one? 153931628204291 (Heat Pump Water Heater)

**Comment by wuwentao** [comment]
this should be a device ID ?  you can check the full debug log and refer to the log that I pasted.

**Comment by hexCut** [comment]
Apologies, I thought you only needed the number. Here’s the full debug string:

`2024-12-18 09:23:46.828 DEBUG (MainThread) [midealocal.discover] Found a supported device: {'device_id': 153931628204291, 'type': 205, 'ip_address': '192.168.20.167', 'port': 6444, 'model': 'RSJ000CB', 'sn': '0000CD311RSJ000CB2618B100053MY44', 'protocol': 3}`

**Comment by wuwentao** [comment]
@hexCut 

add some debug code in this PR: https://github.com/midea-lan/midea-local/pull/345
you can manual change the midealocal python pip lib source code to test with it.

brief step:
1. login to HAOS
2. start homeassitant docker
3. replace file for midealocal python pip package in homeassitant core(use wget to download github PR raw file and replace it)
4. **enable debug log for both midea_ac_lan and midealocal in `configuration.yaml`**
5. **reboot HA.**



how to change it, below is the detail steps:

1. if you are using HAOS, you should install Add-on `[Advanced SSH & Web Terminal]` and disable `[Protected Mode]`, once it done, you can ssh to your HAOS, and enter docker in HAOS `docker exec -it homeassistant /bin/bash`, contiune the step to check `midea-local` install path
2. if you not using HAOS, just check `midea-local` install path as the next step
3.get midea-local install path with command `pip show midea-local`, it will show info for you, for example, my location is Location: /usr/local/lib/python3.12/site-packages
4. then we can replace the file under `/usr/local/lib/python3.12/site-packages/midealocal/devices/cd/` with the code changes, using vi command line tool or wget to download github raw file.
5. make sure you have enabled debug log in `configuration.yaml`
6. reboot HA and test with it, confirm with the result or debug log.

after device reboot 3-5 minutes, download full debug log to your PC and upload this log file to current github issue
I will continue improve it with the debug log output.

in addition, default is `lua_protocol=auto` I have process it for your device with subtype and model, if something can't works, you can set customize option with `lua_protocol=new`.

old device will use `lua_protocol=old`.


this is the first draft version, it may include error, **especially for fahrenheit or celsius switch**, I'm not sure with device response and current result, need to check the detail via debug log.

**Comment by wuwentao** [comment]
@hexCut any updates?

**Comment by hexCut** [comment]
Sorry, I just got back and I didn’t quite understand what exactly I need to do. I’m using this: https://github.com/wuwentao/midea_ac_lan/

**Comment by wuwentao** [comment]
> Sorry, I just got back and I didn’t quite understand what exactly I need to do. I’m using this: https://github.com/wuwentao/midea_ac_lan/

this is a manual test step with source code edit.
I have add all these detail steps, if you can't understand it, just ignore it, as you can't finished these manual tests by yourself.

wish there is any 0xCD device user can do it and confirm the test result.
thanks for the info share.

**Comment by wuwentao** [comment]
please check with latest v0.6.6 release, it should include the fix.
if there is any issue exist, please upload your debug log file, thanks
